catalogd/internal/storage/index.go

```diff
@@ -0,0 +1,133 @@
+package storage
+
+import (
+	"cmp"
+	"encoding/json"
+	"fmt"
+	"io"
+	"slices"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/operator-framework/operator-registry/alpha/declcfg"
+)
+
+type index struct {
```

Comment:
We should add some GoDoc to `index` to help our future selves understand how the indexing works.
```suggestion
// index is a simple index of sections of an FBC file used to lookup FBC blobs that
// match any combination of their schema, package, and name fields. A section is the
// byte offset and length of an FBC blob within the file.
//
// This index strikes a balance between space and performance. It indexes each field
// separately, and performs logical set intersections at lookup time in order to implement
// a multi-parameter query.
//
// Note: it is permissible to change the indexing algorithm later if it is necessary to
// tune the space / performance tradeoff. However care should be taken to ensure
// that the actual content returned by the index remains identical, as users of the index
// may be sensitive to differences introduced by index algorithm changes (e.g. if the
// order of the returned sections changes).

type index struct {
```

Reply:
I realized that we'd want robust test cases for the indexing algorithm too, but that too is probably a follow up in the interest of time.

Reply:
We may have a bunch of coverage of this in the higher level tests, so maybe unnecessary.

Reply:
We do have sufficient coverage with the higher level tests, still writing these tests helped me with grasping the indexing algorithm more https://github.com/operator-framework/operator-controller/pull/1702

ie my argument for the utility of these tests are "better code documentation with unit tests that help understand behavior"